### PR TITLE
CAPI e2e tests

### PR DESCRIPTION
**Pull request by bschimke95**
## Summary
This PR enables and fixes the e2e tests for our CAPI providers. Specifically they test that:
* Workload clusters can be created and correctly form a cluster.
* Clusters can be scaled up and down
* Machines are remediated if they are unhealthy

Right now, the machine upgrade test is disabled as it is more involved. This will be tackled as follow-up work.

## Notable changes

* Adds a `Dockerfile` to build the providers
* Fix api versions for cluster templates
* Move `WaitForNodesReady` from the CAPI test framework to the `helpers` as we only want to test for the minor Kubernetes version while `framework.WaitForNodesReady` does a full version check

### CODE REVIEW

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  build-e2e-images:
+    name: Build & Run E2E Images
+    runs-on: [self-hosted, linux, X64, jammy, xlarge]
+    steps:
+      - name: Check out repo
+        uses: actions/checkout@v4
+      - name: Install requirements
+        run: |
+          sudo apt install make
+          sudo snap install go --classic
```
Comment (eaudetcobello) [inline .github/workflows/e2e.yaml:19]: Let's use an action to be consistent with the rest of the code base and ensure we know which version is installed.

```
- name: Install Go
  uses: actions/setup-go@v5
  with:
    go-version: "1.22"
```

File: .github/workflows/e2e.yaml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+          sudo apt install docker-buildx
+          sudo snap install kubectl --classic
```
Comment (eaudetcobello) [inline .github/workflows/e2e.yaml:21]: Specify a version/channel here.

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          "Workload cluster scaling",
+          "Workload cluster upgrade",
+          ]
+    steps:
+      - name: Check out repo
+        uses: actions/checkout@v4
+      - name: Install requirements
+        run: |
+          sudo apt install make
+          sudo snap install go --classic
```
Comment (eaudetcobello) [inline .github/workflows/e2e.yaml:63]: Idem.

File: .github/workflows/e2e.yaml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+          sudo apt install docker-buildx
+          sudo snap install kubectl --classic
```
Comment (eaudetcobello) [inline .github/workflows/e2e.yaml:65]: Idem.

File: .github/workflows/e2e.yaml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        run: |
+          # Prevents https://cluster-api.sigs.k8s.io/user/troubleshooting#cluster-api-with-docker----too-many-open-files
+          sudo sysctl fs.inotify.max_user_watches=1048576
+          sudo sysctl fs.inotify.max_user_instances=8192
+      - name: Run e2e tests
+        run: |
+          sudo GINKGO_FOCUS="${{ matrix.ginkgo_focus }}" SKIP_RESOURCE_CLEANUP=true make test-e2e
+      - name: Setup tmate session
+        if: ${{ failure() }}
+        uses: canonical/action-tmate@main
```
Comment (eaudetcobello) [inline .github/workflows/e2e.yaml:88]: Are you keeping this for production or was it a debugging thing?

File: test/e2e/helpers.go [hunk 6]
```diff
@@ -587,6 +591,40 @@ func UpgradeControlPlaneAndWaitForUpgrade(ctx context.Context, input UpgradeCont
...
+// and are ready.
+func WaitForNodesReady(ctx context.Context, input WaitForNodesReadyInput) {
+	Eventually(func() (bool, error) {
+		nodeList := &corev1.NodeList{}
+		if err := input.Lister.List(ctx, nodeList); err != nil {
+			return false, err
+		}
+		nodeReadyCount := 0
+		for _, node := range nodeList.Items {
+			n := node
```
Comment (eaudetcobello) [inline test/e2e/helpers.go:611]: Is this needed?

**Review by eaudetcobello — APPROVED** [review]
Nice work

**Review by eaudetcobello — CHANGES_REQUESTED** [review]
Nice work

File: .github/workflows/e2e.yaml (hunk 5 above)
Comment (bschimke95) [reply .github/workflows/e2e.yaml:88]: This was only for debugging. Thanks

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (bschimke95) [reply .github/workflows/e2e.yaml]: I ended up just setting the `channel` for the snap. For some reason the go binary was not available in follow-up steps and I don't think it is worth to dig into that right now.

File: get-docker.sh
Comment (neoaggelos) [inline get-docker.sh:1]: is this used anywhere?

File: .github/workflows/e2e.yaml [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    needs: build-e2e-images
+    strategy:
+      matrix:
+        ginkgo_focus: [
+          "KCP remediation",
+          "MachineDeployment remediation",
+          "Workload cluster creation",
+          "Workload cluster scaling",
+          "Workload cluster upgrade",
+          ]
```
Comment (neoaggelos) [inline .github/workflows/e2e.yaml:57]: not important, but this is a multi-line list already, better go with

```
- KCP remediation
- MachineDeployment remediation
....
```

**Review by neoaggelos — COMMENTED** [review]
let's also update the README.md with instructions for running these locally

**Review by eaudetcobello — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -5,3 +5,20 @@ This repository contains bootstrap and control plane providers to deploy Canonic
...
+```
+
+For the e2e tests, run:
+
+```shell
+make docker-build-e2e  # run only once to build the images
+make test-e2e
+```
+
+Visit the [tests README](./test/e2e/README.md) for more information.
```
Comment (neoaggelos) [inline README.md:24]: Let's keep the readme clean from development instructions. Also, the e2e README.md should be updated, there's outdated references to tilt etc, let's make sure we clean them up and make it a simple straightforward doc that any engineer who works on this can setup their environment, and run the tests

File: test/e2e/helpers.go [hunk 9]
```diff
@@ -596,3 +633,56 @@ func byClusterOptions(name, namespace string) []client.ListOption {
...
+	case "major":
+		return v1[0] == v2[0]
+	case "minor":
+		return v1[0] == v2[0] && v1[1] == v2[1]
+	case "patch":
+		return v1[0] == v2[0] && v1[1] == v2[1] && v1[2] == v2[2]
+	default:
+		return v1[0] == v2[0] && v1[1] == v2[1] && v1[2] == v2[2]
+	}
+}
```
Comment (neoaggelos) [inline test/e2e/helpers.go:671]: use the semver package for this instead

File: test/e2e/e2e_suite_test.go [hunk 10]
```diff
@@ -170,7 +170,7 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	kubeconfigPath := parts[3]
 
 	e2eConfig = loadE2EConfig(configPath)
-	bootstrapClusterProxy = framework.NewClusterProxy("bootstrap", kubeconfigPath, initScheme(), framework.WithMachineLogCollector(framework.DockerLogCollector{}))
+	bootstrapClusterProxy = framework.NewClusterProxy("bootstrap", kubeconfigPath, initScheme(), framework.WithMachineLogCollector(DockerLogCollector{}))
```
Comment (neoaggelos) [inline test/e2e/e2e_suite_test.go:173]: why is this needed?

File: test/e2e/docker_logcollector.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	expv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/infrastructure/container"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kind/pkg/errors"
+)
+
+// DockerLogCollector collect logs from a CAPD workload cluster.
+type DockerLogCollector struct{}
```
Comment (neoaggelos) [inline test/e2e/docker_logcollector.go:39]: mention that DockerLogCollector is `framework.DockerLogCollector` and describe what is different here (for future reference). Will make it easier to keep the code up to date if framework moves along in the meantime

File: test/e2e/e2e_suite_test.go (hunk 10 above)
Comment (bschimke95) [reply test/e2e/e2e_suite_test.go:173]: I used this to also collect pebble logs but actually just ended up debugging the live cluster with tmate. Let's remove it

File: test/e2e/docker_logcollector.go (hunk 11 above)
Comment (bschimke95) [reply test/e2e/docker_logcollector.go]: Removed as we don't need it anymore

**Review by neoaggelos — APPROVED** [review]
